lib/block_view_validator.go

```diff
@@ -0,0 +1,2133 @@
+	}
+	if validatorEntry != nil && // ValidatorEntry exists
+		!validatorEntry.DisableDelegatedStake && // Existing ValidatorEntry.DisableDelegatedStake = false
+		metadata.DisableDelegatedStake { // Updating DisableDelegatedStake = true
+
+		// Retrieve existing StakeEntries.
+		prevStakeEntries, err := bav.GetStakeEntriesForValidatorPKID(transactorPKIDEntry.PKID)
+		if err != nil {
+			return errors.Wrapf(err, "UtxoView.IsValidRegisterAsValidatorMetadata: error retrieving existing StakeEntries: ")
+		}
+
+		// Error if any belong to someone other than the validator.
+		for _, stakeEntry := range prevStakeEntries {
+			if !stakeEntry.StakerPKID.Eq(transactorPKIDEntry.PKID) {
+				return errors.Wrapf(
+					RuleErrorValidatorDisablingExistingDelegatedStakers, "UtxoView.IsValidRegisterAsValidatorMetadata: ",
+				)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) IsValidUnregisterAsValidatorMetadata(transactorPublicKey []byte) error {
+	// Validate ValidatorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return errors.Wrapf(RuleErrorInvalidValidatorPKID, "UtxoView.IsValidUnregisterAsValidatorMetadata: ")
+	}
+
+	// Validate ValidatorEntry exists.
+	validatorEntry, err := bav.GetValidatorByPKID(transactorPKIDEntry.PKID)
+	if err != nil {
+		return errors.Wrapf(err, "UtxoView.IsValidUnregisterAsValidatorMetadata: ")
+	}
+	if validatorEntry == nil || validatorEntry.isDeleted {
+		return errors.Wrapf(RuleErrorValidatorNotFound, "UtxoView.IsValidUnregisterAsValidatorMetadata: ")
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) IsValidUnjailValidatorMetadata(transactorPublicKey []byte) error {
+	// Validate ValidatorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return errors.Wrapf(RuleErrorInvalidValidatorPKID, "UtxoView.IsValidUnjailValidatorMetadata: ")
+	}
+
+	// Validate ValidatorEntry exists.
+	validatorEntry, err := bav.GetValidatorByPKID(transactorPKIDEntry.PKID)
+	if err != nil {
+		return errors.Wrapf(err, "UtxoView.IsValidUnjailValidatorMetadata: ")
+	}
+	if validatorEntry == nil || validatorEntry.isDeleted {
+		return errors.Wrapf(RuleErrorValidatorNotFound, "UtxoView.IsValidUnjailValidatorMetadata: ")
+	}
+
+	// Validate ValidatorEntry is jailed.
+	if validatorEntry.Status() != ValidatorStatusJailed {
+		return errors.Wrapf(RuleErrorUnjailingNonjailedValidator, "UtxoView.IsValidUnjailValidatorMetadata: ")
+	}
+
+	// Retrieve CurrentEpochNumber.
+	currentEpochNumber, err := bav.GetCurrentEpochNumber()
+	if err != nil {
+		return errors.Wrapf(err, "UtxoView.IsValidUnjailValidatorMetadata: error retrieving CurrentEpochNumber: ")
+	}
+
+	// Validate sufficient epochs have elapsed for validator to be unjailed.
+	if validatorEntry.JailedAtEpochNumber+bav.Params.ValidatorJailEpochDuration > currentEpochNumber {
```

Comment:
I wonder if we want to make this UnjailEpochNumber rather than JailedAt. This would allow us the flexibility to jail validators for different amounts of time depending on what they did. Not sure if we want to make this change now or later. Maybe fine to leave as-is and revisit if we have scenarios where we want to jail differently.